### ISSUE TITLE
Why has this library gotten so little attention?

### ISSUE THREAD

**Issue by pha3z** [michaellperry/UpdateControls#16]
I am building my first WPF application -- having come from a background in server development, so the MVVM pattern is new to me.  So far I've been very happy with an MVVM framework I decided upon called Stylet.

I got to a point in my project where I have this dilemma that I need to maintain a list of objects that are displayed in the GUI.  These objects represent program state and are also stored in a database so that whenever the user exits and relaunches the application, the app resumes from where user left off.  As a result, I am storing information about the objects as well as their order that they appear in the UI.

The weirdness here is that the data is clearly reflective of the UI state itself, but I don't want to tie it directly to the WPF implementation.  I decided to abstract it into a model, but then that became overly complex because I have to implement additional plumbing in the viewModel just to interact with the model.  So that got me to start asking questions about what exactly a model really is and the nature of its data.

I read that you shouldn't put ObservableCollections directly in models, which stunned me because in this case I need to maintain data abstracted away from the UI for the sake of code reuse for other platform. 
Because the data is itself directly representative of the UI, it doesn't make any sense to have this secondary layer of observable collection in the viewmodel and all the plumbing code to bind it to another collection in the model.

That's when I found out about UpdateControls and the concept of Independent and Dependent data.  I read about it briefly and thought "wow.  this makes a ton of sense.  why haven't I heard about this before?"  And then I look at the github. I see Michael has been maintaining the project actively, but the userbase is practically non-existant.

So as a developer new to MVVM on this first big desktop app project, I wonder if you can shine any light in the hole for me.  Is there some reason why this library didn't take off when it was introduced many years ago?  Do people just not understand the problem it solves?  Is my particular problem an ideal example of where UpdateControls would shine?

Hope to hear from you. Thanks for taking the time to read this and consider! :)

**Comment by michaellperry** [comment]
I assume that you are referring to my advice in [Common Mistakes While Using ObservableCollection](http://updatecontrols.net/doc/tips/common_mistakes_observablecollection.html) when you say you shouldn't put them directly into models. I think that the reason you found it so surprising on first reading is that most advice you read about MVVM does not make a clear distinction between the model and the view model. Once you make that distinction, you recognize that these should be different sets of types. You would never have a collection of models within a view model. And you would never use a collection type intended for data binding to just store data.

In answer to your overall question, I think perhaps it is this very thing that kept Update Controls from seeing wider adoption. People followed the common MVVM advice and put their models inside of an ObservableCollection in the view model. It worked. They shipped code, and were productive. Why then follow the advice of some guy who says that that is wrong?

As you've experienced, problems do eventually arise from not clearly separating your model and view model layers. However, by the time that happens, you are already invested in your existing framework. You have already learned a set of patterns. It is easier to fix the problem in your code than to change your way of thinking, rearchitect your application around new patterns, and rewrite large swaths of code.

Still, I found that people are inspired by Update Controls (and Assisticant, the rebranding I did around XAML and mobile). You can see the inspiration in Knockout. And you can see how people have explored other solutions to the problem in Angular and React, to varying levels of success. I'll take that as a win.

Thanks. Hope this library or Assisticant serves you well.

**Comment by pha3z** [comment]
Is Assisticant intended as an alternative/replacement for Update Controls?

**Comment by michaellperry** [comment]
It is, unless you are using Windows Forms. I did not port the "controls" of Update Controls over to Assisticant. I also left Silverlight behind. I ported the dependency tracking algorithm and the WPF XAML support. I've added Xamarin mobile support, .NET Standard, and UWP only to Assisticant.

This allowed me to make a clean break from the very thing that gave Update Controls its name. It also allowed me to replace the "mathy" terms (Independent/Dependent) with more industry accepted terms (Observable/Computed). These terms originally came from Knockout, and evoke a comparison with ReactiveUI.

I find that the adoption of Assisticant is better than Update Controls, so it would seem that this was a good choice.

**Comment by pha3z** [comment]
Thank you, Michael!

**Comment by qwertie** [comment]
Hi Michael, when I went to the [UpdateControls web site](http://updatecontrols.net/cs/index.html) I was a bit worried that it was abandoned, since the image at the top-left corner doesn't load and the link to [this video](http://updatecontrols.net/videos/Magic/Magic.html) and others is also broken. Plus there's no link to GitHub for some reason.

Definitely it seemed to me that Update Controls was a trailblazor - it predates MobX and SwiftUI which both implement essentially the same concept (for JavaScript and Swift respectively) but are more popular (edit: I see that KnockoutJS also does dependency tracking... I wonder how old it is).

Back when I used Update Controls, I loved it, with a couple of exceptions, both performance-related:

1. It doesn't understand lists. A list is handled the same way as a single datum, so if you've got an `IndependentList<Foo>` and a `DependentList<FooViewModel>` and you add a single item to the IndependentList, the entire DependentList is recomputed from scratch, as well as anything else that depends on it.
2. It doesn't understand changes that have no effect. Suppose I have a `Dependent<int> C` that computes the maximum value of two other precedents A and B, and then I decrease the smaller precedent A. In this case C does not change, but anything that depends on C is updated anyway. Not only are the dependents updated, but there is no way to suppress this behavior.

Have you been able to address these problems in the last few years, Michael?

I still love UpdateControls, but recently I've been working on pre-existing **huge** software with a pre-existing commitment to the old-fashioned (inferior) way of doing things. And to be honest, without solutions to the performance issues, it is hard to recommend Update Controls with the necessary enthusiasm. And on some products we've got web front ends that aren't C# anyway. And sadly, no one is using MobX.

**Comment by michaellperry** [comment]
Hey, @qwertie! Good to hear from you again. It has been a while.

The website is in disrepair. I'm a bit sad to see what I've let happen as well. I was able to fix the broken image easily enough, but the videos are probably lost to time. For now, I've removed primary links to the videos page, but it's still in the blog posts. I had scraped the content from the old WordPress site, so fixing that is a bit of a mess. Anybody who would like to help me fix up the site can find it at [updatecontrols_net](https://github.com/michaellperry/updatecontrols_net).

I've shifted my focus over to Assisticant. It has the same core algorithm, but more approachable class names, and better XAML support. Following the popularity of Knockout, I adopted Observable and Computed rather than the mathematically-based Independent and Dependent. Perhaps this contributed to a higher adoption of Assisticant than Update Controls.

And you are correct about the timeline. Update Controls predates Knockout (and in fact was the inspiration, according to Steve Sanderson), but the Knockout naming inspired Assisticant. I'm happy to have played some small indirect part in seeding the research that eventually brought us things like React.

As for the performance issues, my usual answer to the first one is to create lists of objects which themselves have Independent fields. That way you don't modify the list to simply change one property. Furthermore, by implementing Equals and GetHashCode, you can take advantage of object recycling when you *do* change the list to minimize the impact on the UI. I haven't been able to come up with a more performant solution that still generalizes.

For the second one, I don't have a good answer. The reason for the problem is that indirect Dependents pull updates forward through all of the intermediate layers. On the one hand, this is a desirable feature, because it allows non-visible Dependents to remain out-of-date for as long as possible. But on the other hand, it means that visible Dependents need to be notified, even if it turns out that an intermediate will *not* actually change value.

In the end, I make the same claim as the React team does. The library minimizes unnecessary UI changes, which is the most costly part of the update. In React, they have the virtual DOM. In Update Controls, we have object recycling. React has opted to give developers *some* control with things like `ComponentShouldUpdate`, whereas I have not. I value generalization over performance. I've avoided giving developers tools that introduce subtle defects, rather than trying to educate them about how to use them properly.

**Comment by qwertie** [comment]
Remember that the first scenario is to _add_ something to the list, so that having independents in the individual objects would not avoid rebuilding the derived list. (I'm not sure if you saw that in the initial post, which I immediately edited.) VMs can be recycled, but recycling is still O(N) in the list size.

I remember coming up with a solution to problem 2 at some point. Let's see if I can remember....

IIRC, Update Controls is based on a directed graph pointing from precedents to dependents (computeds?). When you update an Independent (observable), Update Controls recursively tells dependents that they are invalid, and deletes edges pointing to them.

But must we delete edges outright? It seems like deleting the immediate edges is sound, as certainly immediate dependents must be recomputed. But for all indirectly-reachable nodes, let's add a measurement called "uncertainty". Example:

~~~
    A (Independent)
    |
    B (Dependent)
    |
    C (Dependent)
    |
    D (Dependent)
~~~

You update A and this causes it to delete the edge(s) to its child(ren) {B}, but other edges are not deleted. Instead, children are marked as "uncertain" recursively:

~~~
    * = invalid (unknown)
    1 = level of uncertainty

    A (Independent)
    
    B*1 (Dependent)
    | 
    C1 (Dependent)
    | 
    D1 (Dependent)
~~~

When you ask for a new value of B, it must first delete _incoming_ edges but there aren't any in this example, and then it recomputes itself. If the new value is different, it deletes the edges to its children and then adjusts the uncertainty level of all children recursively: the uncertainty associated with B is removed recursively, but uncertainty is added to C recursively, for a net change of zero. Also, immediate children are marked as invalid.

~~~
    A (Independent)
    
    B (Dependent)
    
    C*1 (Dependent)
    | 
    D1 (Dependent)
~~~

If instead B had not changed, the uncertainty associated with B and its dependents could be decremented, leaving children valid so that they don't have to be updated:

~~~
    A (Independent)
    |
    B (Dependent)
    |
    C (Dependent)
    |
    D (Dependent)
~~~

If instead the value of D had been requested, D is merely uncertain, not invalid, so it must tell C to validate, which in turn causes B to validate. After doing this, D should be either invalid or certain (uncertainty=0).

Notice that "computation" on D is not really eliminated, since it has to do the work of walking the dependency tree, but without solving problem #1 (and I haven't quite solved problem #1) this is still better, because "full" recomputations are more likely to be expensive.

Okay let's try a more complicated one, because I'm not a math guy (I don't do proofs, I just come up with something harder and reason through it):

~~~
       A     B   C
       |\   /|\  |
       | \ / | \ |
       |  D  |  \|
       | /   |   E
       |/    |   |
       F     |   |
      / \   /    |
     /   \ /    /
    G     H----/
      \   |   /
       \  |  /
        \ | /
         \|/
          I
~~~
        
Let's say we update A. Then the graph becomes 

~~~
    * invalid
    1,2,3,4 level of uncertainty

       A     B   C
            /|\  |
           / | \ |
          D*1|  \|
         /   |   E
        /    |   |
       F*2   |   |
      / \   /    |
     /   \ /    /
    G2    H2---/
      \   |   /
       \  |  /
        \ | /
         \|/
          I4
~~~

Here I'm thinking F itself is uncertain but it also has an uncertain precedent, for a total uncertainty of 2, with the uncertainty of I's two precedents summing.
          
Suppose F's value is requested. F depends on A and D, but in general we don't know what it depends on, so we must start by deleting the edge DF and reconstruct the dependency set. This implies extra memory is needed to store a list of precedents at each dependent (but maybe this can be optimized somehow). In addition, the uncertainty on F should decrease by 1 recursively as a result of deleting that link (but I think this operation can be deferred until later).

So, F gets A and D. D's getter recomputes D, but again, its precedent set is not guaranteed to be A and B, so it must start by deleting the remaining edge BD.

It doesn't really matter if D is the same as before, since D has no child edges yet.

Next, F recomputes. If F is the same as before, it'll subtract 1 from the uncertainty of itself and its children, but actually 2 is subtracted due to the earlier deletion of DF, producing a pristine result:

~~~
       A     B   C
       |\   /|\  |
       | \ / | \ |
       |  D  |  \|
       | /   |   E
       |/    |   |
       F     |   |
      / \   /    |
     /   \ /    /
    G     H----/
      \   |   /
       \  |  /
        \ | /
         \|/
          I
~~~

If F changes, on the other hand, links to its children disappear, and uncertainty of its former children is decreased by 2 and increased by 1:

~~~
       A     B   C
       |\   /|\  |
       | \ / | \ |
       |  D  |  \|
       | /   |   E
       |/    |   |
       F     |   |
            /    |
           /    /
    G*1   H*1--/
      \   |   /
       \  |  /
        \ | /
         \|/
          I2
~~~

Now let's say C changes and we get this result...

~~~
       A     B   C
       |\   /|    
       | \ / |    
       |  D  |    
       | /   |   E*
       |/    |   |
       F     |   |
            /    |
           /    /
    G*1   H*2--/
      \   |   /
       \  |  /
        \ | /
         \|/
          I4
~~~

Next let's get `I`. Since I is uncertain, it asks its uncertain precedents G, H and E to validate and they must all recompute. G and H both reestablish their link to F (though H's link to B and E must be deleted and recreated). G and H and E will (in some order where E comes before H) either invalidate I or reduce its uncertainty. As usual, if I is invalidated it must be recomputed, otherwise not.

I can't remember what solution I came up with a few years back, but I'm sure this is not it.

Assuming this approach is sound, I can see that the naive implementation - in which you actually use recursive calls to increment or decrement uncertainty - would be unacceptably slow for certain graphs like this one:

~~~
      AA
      /\
     /  \
    A    B
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    C    D
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    E    F
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    G    H
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    I    J
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    K    L
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    M    N
~~~

This is because when you invalidate AA, uncertainty piles up exponentially, which means the number of naive method calls would pile up likewise:

~~~
      AA
        
         
    A*1  B*1
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    C2   D2
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    E4   F4
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    G8   H8
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    I16  J16
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    K32  L32
    |\  /|
    | \/ |
    | /\ |
    |/  \|
    M64  N64
~~~

This can probably be optimized somehow, though I wonder if overflowing `int` is a concern. Anyway... food for further thought.

**Comment by michaellperry** [comment]
I think your solution generalizes nicely. I hadn't been able to find one that does. The level of uncertainty is a proxy for the upstream subgraph. It gives you information without performing the traversal. It seems there should be a different proxy that does not fall to the exponential growth problem. That's worth a look.

Edges are removed in order to make dependencies dynamic. But as you point out, they could be removed just before computation and achieve the same effect. The benefit of late removal is that the graph is preserved downstream in the case of a non-changing update.

The cost of late removal is that additional changes before the next update cycle incur extra traversals. Precedents have to notify their dependents along edges that would have been destroyed. That traversal now modifies the level of uncertainty, or whatever other proxy might replace it.

These competing optimizations favor different scenarios. Late removal favors non-changing updates. Early removal favors changing updates. I made the assumption that most updates are changing. I don't know if that's true in practice.

In particular, we know that lists (and even more dictionaries) present mostly-non-changing updates. Perhaps there is a way to preserve the majority of a graph for collections, even if we continue with early removal for other types. I'm thinking about the patterns used in reactive programming.

This is good stuff. I don't want to lose it. Would you mind creating a new issue?
